src/SB/Core/x/xParEmitter.cpp

```diff
@@ -1,3 +1,828 @@
+            val = r->val[0];
+        }
+        break;
+    case 4:
+        val = isin((1.0 * (val * r->oofreq)));
+        val = ((r->val[1] - val_0) * (1.0 * val + 1.0) + val_0);
+        break;
+    case 5:
+        val = icos((1.0 * (val * r->oofreq)));
+        val = ((r->val[1] - val_0) * (1.0 * val + 1.0) + val_0);
+        break;
+    case 7:
+        if ((val * r->freq) < 0.0)
+        {
+            val = r->val[0];
+        }
+        else
+        {
+            val = r->val[1];
+        }
+    }
+    return val;
+}
+
+xPar* xParEmitterEmitSetTexIdxs(xPar* p, xParSys* ps)
+{
+    U8 ps_birthMode;
+    xParCmdTex* ps_cmdTex;
+
+    ps_cmdTex = ps->group->m_cmdTex;
+    if (ps_cmdTex == NULL)
+    {
+        return NULL;
+    }
+    ps_birthMode = ps_cmdTex->birthMode;
+    if ((S32)ps_birthMode != 1)
+    {
+        if ((S32)ps_birthMode < 1)
+        {
+            if ((S32)ps_birthMode < 0)
+            {
+            }
+            else
+            {
+                p->m_texIdx[0] = 0;
+                p->m_texIdx[1] = 0;
+            }
+        }
+        else if ((S32)ps_birthMode < 7)
+        {
+        }
+    }
+    else
+    {
+        p->m_texIdx[0] = (xrand() >> 0x11U) % ps_cmdTex->cols;
+        p->m_texIdx[1] = (xrand() >> 0x11U) % ps_cmdTex->rows;
+    }
+    return p;
+}
+
+xPar* xParEmitterEmit(xParEmitter* pe, F32 emit_dt, F32 par_dt)
+{
+    xPar* p;
+    xPar* last_p;
+    xMat4x3* bone_mat;
+    xParSys* ps;
+    xParEmitterPropsAsset* prop;
+    xParEmitterAsset* pea;
+    xVec3 emitPosition;
+    xVec3 bone_vel;
+    xGroup* g;
+    xEnt* attach_ent;
+
+    F32 life;
+    F32 size_birth;
+    F32 size_death;
+
+    S32 i;
+    S32 c;
+    S32 attachGroupTotal;
+    S32 rate_has_elapsed;
+    U32 attachGroupIndex;
+    S32 count;
+    S32 emitAgain;
+    bool marker;
+    S32 group_ptr_chk;
+    U32 temp_maxPar;
+
+    if (pe->parSys == 0)
+    {
+        p = (xPar*)0x0;
+    }
+    else
+    {
+        p = (xPar*)0x0;
+        pea = pe->tasset;
+        prop = pe->prop;
+        pe->rate_time = pe->rate_time + emit_dt;
+        size_birth = pe->rate_time;
+        if (pe->rate_time == 0.0)
+        {
+            pe->rate_time = 0.0;
+        }
+        while (pe->rate_time > prop->value[0].freq)
+        {
+            pe->rate_time = pe->rate_time - prop->value[0].freq;
+        }
+        size_birth = xParInterpCompute(pe->rate_mode, prop->value, pe->rate_time, 0, 0.0);
+        pe->rate = size_birth;
+        pe->rate_fraction = pe->rate_fraction + ((double)size_birth * emit_dt);
+        pe->rate_fraction_cull = pe->rate_fraction_cull + (F32)((double)size_birth * emit_dt);
+        size_birth = floorf(pe->rate_fraction);
+        rate_has_elapsed = size_birth;
+        if (rate_has_elapsed > 0)
+        {
+            pe->rate_fraction = pe->rate_fraction - rate_has_elapsed;
+        }
+        if (rate_has_elapsed == 0)
+        {
+            p = (xPar*)0x0;
+        }
+        else
+        {
```

Comment:
Please add a `// TODO` here for cleaning up the logic - works for now but I don't think it needs to be this nested.